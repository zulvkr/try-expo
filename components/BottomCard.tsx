import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { View, ViewProps } from './Themed'
import { StyleSheet } from 'react-native'

interface BottomCardProps extends ViewProps {
  children: React.ReactNode
}

export default function BottomCard({ children, ...props }: BottomCardProps) {
  const insets = useSafeAreaInsets()

  return (
    <View
      {...props}
      style={[
        styles.container,
        {
          paddingBottom: insets.bottom + 20
        }
      ]}
    >
      {children}
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    padding: 20,
    elevation: 5,
    // ios shadow
    shadowColor: '#000',
    shadowOffset: { width: 0, height: -2 },
    shadowOpacity: 0.23,
    shadowRadius: 2.62
  }
})
