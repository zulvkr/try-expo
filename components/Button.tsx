import {
  ActivityIndicator,
  Pressable,
  PressableProps,
  View,
  ViewStyle
} from 'react-native'
import { Text, useThemeColor } from './Themed'
import Colors from '@/constants/Colors'
import { Ref, forwardRef } from 'react'

export type ButtonProps = Omit<
  PressableProps & { children: React.ReactNode },
  'style'
> & {
  variant?: 'filled' | 'outlined' | 'tonal' | 'text'
} & { style?: ViewStyle } & {
  renderLeft?: () => React.ReactNode
  renderRight?: () => React.ReactNode
} & {
  disabled?: boolean
  loading?: boolean
}

type keyOfColor = keyof typeof Colors.light & keyof typeof Colors.dark

type buttonColorConfig = {
  background: keyOfColor
  text: keyOfColor
  border: keyOfColor
}

const buttonColorConfig: Record<string, buttonColorConfig> = {
  filled: {
    text: 'filledButtonTextColor',
    background: 'filledButtonColor',
    border: 'background'
  },
  outlined: {
    text: 'outlineButtonColor',
    border: 'outlineButtonColor',
    background: 'background'
  },
  tonal: {
    text: 'tonalButtonTextColor',
    background: 'tonalButtonColor',
    border: 'background'
  },
  disabled: {
    text: 'textSecondary',
    background: 'textInputBackground',
    border: 'background'
  }
}

export const Button = forwardRef(function Button(
  props: ButtonProps,
  ref: Ref<View>
) {
  const { variant = 'filled', style, ...otherProps } = props

  const backgroundThemeKey =
    buttonColorConfig[props.disabled ? 'disabled' : variant].background
  const textThemeKey = buttonColorConfig[variant].text
  const borderThemeKey = buttonColorConfig[variant].border

  const backgroundColor = useThemeColor({}, backgroundThemeKey)
  const textColor = useThemeColor({}, textThemeKey)
  const borderColor = useThemeColor({}, borderThemeKey)

  return (
    <Pressable
      disabled={props.loading || props.disabled}
      ref={ref}
      style={({ pressed }) => ({
        opacity: pressed ? 0.5 : 1,
        padding: 15,
        paddingVertical: 18,
        borderRadius: 99,
        alignItems: 'center',
        borderWidth: variant === 'outlined' ? 1 : 0,
        borderColor,
        backgroundColor,
        ...(style ?? {})
      })}
      {...otherProps}
    >
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'center',
          minHeight: 22,
          alignItems: 'center'
        }}
      >
        {props.renderLeft?.()}
        {props.loading ? (
          <ActivityIndicator size={22} color={textColor} />
        ) : (
          <Text style={{ fontWeight: 'bold', color: textColor }}>
            {props.children}
          </Text>
        )}
        {props.renderRight?.()}
      </View>
    </Pressable>
  )
})

export type IconButtonProps = Omit<
  PressableProps & { children: React.ReactNode },
  'style'
> & { style?: ViewStyle }

export const IconButton = forwardRef(function IconButton(
  props: ButtonProps,
  ref: Ref<View>
) {
  const { style, ...otherProps } = props

  return (
    <Pressable
      ref={ref}
      style={({ pressed }) => ({
        opacity: pressed ? 0.5 : 1,
        height: 32,
        width: 32,
        borderRadius: 16,
        alignItems: 'center',
        justifyContent: 'center',
        ...(style ?? {})
      })}
      {...otherProps}
    >
      <Text
        style={{
          flexDirection: 'row',
          justifyContent: 'center',
          alignItems: 'center'
        }}
      >
        {props.children}
      </Text>
    </Pressable>
  )
})
