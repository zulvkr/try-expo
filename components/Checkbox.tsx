import { MaterialCommunityIcons } from '@expo/vector-icons'
import { Text, useThemeColor } from './Themed'
import { Pressable, StyleSheet, ViewProps } from 'react-native'
import { IconProps } from '@expo/vector-icons/build/createIconSet'
import { useState } from 'react'
import Colors from '@/constants/Colors'

export type CheckboxProps = Omit<IconProps<''>, 'name' | 'color'> & {
  value?: boolean
  onChange?: (value: boolean) => void
  color?: (isChecked: boolean) => string
  label?: string
  containerProps?: ViewProps
}

export default function Checkbox(props: CheckboxProps) {
  const { value, onChange, color, label, containerProps, ...otherProps } = props

  const [isChecked, setIsChecked] = useState(value ?? false)

  const handlePress = () => {
    setIsChecked(!isChecked)
    onChange?.(!isChecked)
  }

  const inactiveCheckboxColor = useThemeColor(
    {
      dark: Colors.dark.tint
    },
    'textSecondary'
  )
  const activeCheckboxColor = useThemeColor({}, 'tint')

  const _value = value !== undefined ? value : isChecked

  const defaultColor = _value ? activeCheckboxColor : inactiveCheckboxColor

  return (
    <Pressable
      onPress={handlePress}
      style={styles.container}
      {...props.containerProps}
    >
      <MaterialCommunityIcons
        name={_value ? 'checkbox-marked' : 'checkbox-blank-outline'}
        color={props.color ? props.color(_value) : defaultColor}
        size={28}
        {...otherProps}
      />
      {props.label && <Text style={styles.checkboxLabel}>{props.label}</Text>}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  checkboxLabel: {
    marginLeft: 16,
    fontSize: 12
  }
})
