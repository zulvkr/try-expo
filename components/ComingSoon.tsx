import { Text, View } from './Themed'

export default function ComingSoon() {
  return (
    <View
      style={{
        // flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        height: 300
        // width: '100%'
      }}
    >
      <Text style={{ fontSize: 32, fontWeight: 'bold' }}>Coming Soon</Text>
    </View>
  )
}
