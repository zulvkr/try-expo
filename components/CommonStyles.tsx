import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  sectionHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 20
  },
  sectionHeaderTitle: {
    fontSize: 16,
    fontWeight: 'bold'
  }
})
