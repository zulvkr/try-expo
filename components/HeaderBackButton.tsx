import { HeaderBackButtonProps } from '@react-navigation/elements'
import { useRouter } from 'expo-router'
import { Image, Pressable, useColorScheme } from 'react-native'

export const customHeaderBackButton = (scheme?: 'dark' | 'light') => {
  return (props: HeaderBackButtonProps) => {
    const globalScheme = useColorScheme()
    const router = useRouter()
    const _scheme = scheme || globalScheme
    return (
      <Pressable
        style={{
          zIndex: 10,
          width: 32,
          height: 32
        }}
        onPress={() => {
          router.back()
        }}
      >
        <Image
          style={{
            width: 32,
            height: 32
          }}
          source={
            _scheme === 'dark'
              ? require('@/assets/images/back-arrow-light-icon.png')
              : require('@/assets/images/back-arrow-dark-icon.png')
          }
        />
      </Pressable>
    )
  }
}

const HeaderBackButton = customHeaderBackButton()

export default HeaderBackButton
