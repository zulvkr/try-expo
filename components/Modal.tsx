import React from 'react'
import {
  Modal as BaseModal,
  View,
  StyleSheet,
  TouchableOpacity,
  Image
} from 'react-native'
import { BlurView } from 'expo-blur'
import { useThemeColor } from './Themed'

export interface ModalProps {
  visible: boolean
  onClose: () => void
  children?: React.ReactNode
  closeOnOverlayClick?: boolean
}

const Modal = ({ visible, onClose, children }: ModalProps) => {
  const backgroundColor = useThemeColor({}, 'background')
  return (
    <BaseModal visible={visible} transparent animationType='fade'>
      <TouchableOpacity
        activeOpacity={0.75}
        style={styles.overlay}
        onPress={onClose}
      >
        <BlurView
          style={[StyleSheet.absoluteFill, styles.blur]}
          intensity={100}
          tint='dark'
        />
        <View style={[{ backgroundColor }, styles.modal]}>{children}</View>
      </TouchableOpacity>
    </BaseModal>
  )
}

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  blur: {},
  modal: {
    width: 300,
    padding: 28,
    borderRadius: 16
  }
})

export interface SuccessModalProps extends ModalProps {}

export const SuccessModal = (props: SuccessModalProps) => {
  return (
    <Modal {...props}>
      <Image
        style={successModalStyles.image}
        resizeMode='contain'
        source={require('@/assets/images/success-party.png')}
      />
      {props.children}
    </Modal>
  )
}

const successModalStyles = StyleSheet.create({
  image: {
    height: 160,
    width: '100%'
  }
})

export default Modal
