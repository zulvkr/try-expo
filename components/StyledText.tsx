import { Ref, forwardRef } from 'react'
import { Text, TextProps, useThemeColor } from './Themed'

export function MonoText(props: TextProps) {
  return <Text {...props} style={[props.style, { fontFamily: 'SpaceMono' }]} />
}

// export function SecondaryText(props: TextProps) {
//   const color = useThemeColor({}, 'textSecondary')
//   return <Text {...props} style={[props.style, { fontFamily: 'SpaceMono' }]} />
// }

export const SecondaryText = forwardRef((props: TextProps, ref: Ref<Text>) => {
  const { style, lightColor, darkColor, ...otherProps } = props
  const color = useThemeColor(
    { light: lightColor, dark: darkColor },
    'textSecondary'
  )

  return <Text {...otherProps} ref={ref} style={[{ color }, style]} />
})
