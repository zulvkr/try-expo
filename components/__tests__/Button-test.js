import { Button } from '@/components/Button'
import renderer from 'react-test-renderer'

describe(`<Button children="Hello, World!" />`, () => {
  it('should render without crashing', () => {
    const tree = renderer.create(<Button children='Hello, World!' />).toJSON()
    expect(tree.children[0].children[0].children[0]).toBe('Hello, World!')
  })
})

describe(`<Button children="Hello, World!" />`, () => {
  it('should has correct color in light theme', () => {
    const tree = renderer.create(<Button children='Hello, World!' />).toJSON()
    expect(tree.props.style.backgroundColor).toBe('#B80C93')
  })
})
