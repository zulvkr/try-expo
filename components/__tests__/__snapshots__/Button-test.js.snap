// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button children="Hello, World!" /> should render without crashing 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#B80C93",
      "borderColor": "#fff",
      "borderRadius": 99,
      "borderWidth": 0,
      "opacity": 1,
      "padding": 15,
      "paddingVertical": 18,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 22,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
          },
          {
            "color": "#fff",
            "fontWeight": "bold",
          },
        ]
      }
    >
      Hello, World!
    </Text>
  </View>
</View>
`;
